src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -830,6 +857,81 @@ evalApp γ ctx e0 es et
       , f == "Language.Haskell.Liquid.ProofCombinators.pleUnfold"
       = arg
       | otherwise = e
+-- **ETA**: PLE before the patch to perform eta expansion was only able to unfold
+-- fully applied functions, by eta equivalence we mean the property that for all
+-- functions f the terms f and (\x -> f x) have the same semantics, we leverage 
+-- this fact to fully apply functions with extra function arguments so to trigger
+-- the standard PLE espansion, see the following example:
+-- We have a function f:
+-- define f (x : int) : int = {(x)}
+-- And we need to prove some constraint of this shape
+-- { f = something }
+-- The original PLE cannot do anything on f as it's not fully applied, instead
+-- if we perform eta expansion on f we obtain the following constraint
+-- { \y:Int -> f y = something }
+-- And now PLE can perform the unfolding of f as it's now fully saturated
+-- { \y:Int -> y = something }
+-- Clearly we need the higerorder flag active as we are generating lambda in
+-- the constraints.
+evalApp γ _ctx e0 es _et
+  | EVar f <- dropECst e0
+  , Just eq <- Map.lookup f (knAms γ)
+  = do
+    isHOAllowed <- gets hoFlag
+    if isHOAllowed then do
```

Comment:
The slowdown in `ple/pos/Permutations.hs` makes me wonder if extensionality shouldn't be required here as well.

Also, the context in which `f` appears seems to be relevant. Eta expansion makes sense in `C f == C g`, or `map f xs == map g xs`, but not in `map f [] == []`. What we are giving up by eta expanding unconditionally is the size of the PLE dumps. But I don't know if there is a good criteria to recognize the relevant contexts.

In the absence of such criteria, I think it might be preferable to have a flag to turn it on.

Reply:
Should we hide it behind the extensionality flag or make a new one? I agree with your point but I can't think of a way to check if the eta expansion is actually needed or not

Reply:
The thing is that we get the eta expansion working 100% (will require some work on lambdas) then everything that is done by extensionality can be done by the eta expansion alone